Enhance search enabled configuration

jraddaoui opened the pull request:
Allow to partially disable/enable Elasticsearch indexing. Extend the
current `search_enabled` config. var. behavior in all the components
to allow a string with a list of enabled search indexes separated by
comma but maintain the current boolean functionality. Available options
for the list are `transfers`, for the Transfer related indexes and
`aips` for the AIPs related indexes. Check the install README files
from each component for more info.

Enhance config. mapping definition to allow a 'process_function' field
where a callback can be passed to allow a more complex parsing process
for specific attributes. Those callbacks must accept the current 
'appconfig.Config' object and the attribute section string.

**Questions:**
This is currently only extending the existing functionality, but I wonder:

- Should we add a note somewhere indicating that the `search_enabled` configuration should be consistent in the three components?
- Should we try to change this multiple declaration at some point and use some global settings for the three components?
- Should this setting be checked somewhere else? For example, showing a warning in the [rebuild commands](https://github.com/artefactual/archivematica/tree/stable/1.7.x/src/dashboard/src/main/management/commands). I can double check this in #1171 and #1173.
- Is this config/env var documented somewhere else besides the install readme files?
- Since we are maintaining backwards compatibility, I guess we don't need to change the declaration of this config/env var in other repositories, but let me know if I'm wrong.

Connects to #1172.

sevein commented on src/MCPClient/lib/settings/common.py:
```diff
@@ -38,8 +38,9 @@
     'elasticsearch_server': {'section': 'MCPClient', 'option': 'elasticsearchServer', 'type': 'string'},
     'elasticsearch_timeout': {'section': 'MCPClient', 'option': 'elasticsearchTimeout', 'type': 'float'},
     'search_enabled': [
-        {'section': 'MCPClient', 'option': 'disableElasticsearchIndexing', 'type': 'iboolean'},
```
This was on purpose, now I remember. For backward compatibility reasons. We made a release at some point where `disableElasticsearchIndexing` was the name of the attribute so now we're committed to support it. This was only true in MCPClient, that's why you don't see it in other components.

sevein commented:
> Should we add a note somewhere indicating that the search_enabled configuration should be consistent in the three components?

That's a good point. I think we should. There is a section in archivematica-docs dedicated to searches (see related PR: https://github.com/artefactual/archivematica-docs/pull/68).

> Should we try to change this multiple declaration at some point and use some global settings for the three components?

Yes, that's something we should do. We need some kind of global application configuration. We're aware of the problem but I don't think we've filed a bug yet!

> Should this setting be checked somewhere else? For example, showing a warning in the rebuild commands. I can double check this in #1171 and #1173.

From the commands you could tell if the config is disabled in the Dashboard so you could show a warning if you have time, I agree that'd be nice.

> Is this config/env var documented somewhere else besides the install readme files? Since we are maintaining backwards compatibility, I guess we don't need to change the declaration of this config/env var in other repositories, but let me know if I'm wrong.

I think you're right. We should be safe without further changes. Good docs in archivematica-docs extending that PR I've linked above should be enough.

sevein commented on src/archivematicaCommon/lib/appconfig.py:
```diff
@@ -39,7 +41,10 @@ def get(self, attr, default=None):
-            return self.get_from_opts_list(attr, attr_opts, default=default)
+            value = self._get_from_opts_list(attr, attr_opts, default=default)
+            if attr == 'search_enabled':
+                value = self._normalize_search_enabled(value)
```
Would it be possible if you have time to update this solution so it's not specific to given attributes? E.g. we could have a new `list` type along with `allowed_values`:

    {'section': 'MCPServer', 'option': 'search_enabled', 'type': 'list', 'allowed_values': ['aips', 'transfers']},

I'd like `appconfig.Config` to be agnostic to specific configurations as much as possible. Would that be possible? Am I asking too much? I'm not sure how much extra work would that require at this point.

jraddaoui replied:
Thanks @sevein!

The problem with this particular attribute is that it should allow boolean values alongside the list and always return a set. I could try to define the normalization rules in the config mapping at the attribute level alongside the options, but that will add some complexity to the mapping and normalization code. Considering that all this (the multi-option and normalization) looks to be done only for this attribute at this point I'm not sure if it worths the effort, but I'll give it a quick try ;)

jraddaoui replied:
How about something like this for the config mapping:

```python
CONFIG_MAPPING = {
    'search_enabled': {
        'section': 'Dashboard',
        'options': [
            {'name': 'disable_search_indexing', 'type': 'iboolean'},
            {'name': 'search_enabled', 'type': 'boolean'},
            {'name': 'search_enabled', 'type': 'list', 'allowed_values': set(['aips', 'transfers'])},
        ],
        'normalize': [
            {
                'type': bool,
                'rules': [
                    {'value': True, 'output': set(['aips', 'transfers'])},
                    {'value': False, 'output': set()},
                ],
            },
        ],
    },
}
```

sevein replied:
I like it but I recognise it's becoming too complicated as you anticipated.

I wonder if we should step back and return to your original suggestion. Perhaps we could teach `Config` how to deal with configs that have a `filter` callable. It'd be like what you did excepting that it's up to the user to pass a callable where the logic specific to a particular configuration attribute is kept, e.g.:

```python
from app_config import Config, process_search_enabled

CONFIG_MAPPING = {
    'search_enabled': {
        'section': 'Dashboard',
        'filter': process_search_enabled,
   }
}
```

`Config` would call the callable in `filter` when available otherwise fallback to the default behaviour (i.e. `options`). The callable would receive as a parameter the instance of `Config` so it can do with it whatever it wants.

sevein reviewed (commented):
Looking good, thanks @jraddaoui. Only one comment about the changes in `Config`.

sevein commented on src/archivematicaCommon/lib/appconfig.py:
```diff
@@ -40,6 +40,8 @@ def get(self, attr, default=None):
         if isinstance(attr_opts, list):
             return self.get_from_opts_list(attr, attr_opts, default=default)
+        if all(k in attr_opts for k in ('section', 'process_function')):
+            return attr_opts['process_function'](self, attr_opts['section'])
```
:clap:

sevein reviewed (approved):
Thank you so much!

jraddaoui committed:
Enhance search enabled configuration

Allow to partially disable/enable Elasticsearch indexing. Extend the
current `search_enabled` config. var. behavior in all the components
to allow a string with a list of enabled search indexes separated by
comma but maintain the current boolean functionality. Available options
for the list are `transfers`, for the Transfer related indexes and
`aips` for the AIPs related indexes. Check the install README files
from each component for more info.

Enhance config. mapping definition to allow a 'process_function' field
where a callback can be passed to allow a more complex parsing process
for specific attributes. Those callbacks must accept the current
'appconfig.Config' object and the attribute section string.

sallain commented:
@sevein does this resolve #1173?

jraddaoui commented:
Hi @sara, it does not. I'll work on that one alongside #1171 in the near future.

jraddaoui commented:
Sorry, @sallain.
